### PR TITLE
Add starter .greptile/ config for DALI reviews

### PR DESCRIPTION
**Pull request by jantonguirao**
## Summary

Adds a starter `.greptile/` configuration to shape Greptile's automated reviews of DALI PRs along the lines of the manual maintainer review style.

- `.greptile/config.json` — 55 structured rules with stable IDs, per-language `scope` globs, and tiered severity (`high` for correctness/safety, `medium` for design/perf, `low` for style).
- `.greptile/rules.md` — free-form review philosophy: read full files, trace callers/callees, distinguish pre-existing from new bugs, use the `[Critical]/[Bug]/[Perf]/[Style]/[Nit]/[Question]` prefix convention.

Rules were distilled from ~8,500 historical maintainer review comments across nine reviewers. Coverage areas:
- CUDA correctness & perf (bounds guards, `__restrict__`, `size_t` index math, scratchpad over `cudaMalloc`, explicit streams, kernel-launch error handling)
- C++ safety (RAII, `assert` vs `DALI_ENFORCE`, `static_cast` over C-style, `make_unique`, self-assignment safety, no `static` in headers without `inline`)
- Python idioms (list-of-bool truthiness, `assert` disabled with `-O`, `@pipeline_def`, `zip(strict=True)`, positive-form booleans)
- API & naming (`Is*`/`Has*` for booleans, enums over magic strings, DALI utility helpers)
- Error messages (named context, expected vs. actual, `make_string`)
- Testing standards (reference comparisons, edge cases, deterministic seeds, no silent skips)
- Docs (copyright year, `@param` name match, RST backtick discipline, no TODO at merge)
- Style (no commented-out code, no debug prints, no reformat-only churn, typo list)

`strictness: 1` matches the verbose review posture; `commentTypes` drops `info` to reduce filler. `ignorePatterns` excludes `third_party/`, generated `_pb2` files, and vendored `_autograph/`. `context.repos: ["NVIDIA/DALI_extra"]` so Greptile can resolve test-data references.

Opened as draft so the rule set can be tuned before enabling. Per-directory configs can later disable inherited rules via `disabledRules` (e.g. relax style rules under `docs/examples/`).

## Test plan

- [ ] Eyeball each rule for false-positive risk; mark noisy ones `"enabled": false`
- [ ] Verify `scope` globs match the actual repo layout (especially `dali/kernels/`)
- [ ] Watch Greptile's behavior on the next 2–3 PRs after merge; iterate
- [ ] Decide whether `dali/test/.greptile/config.json` and `docs/examples/.greptile/config.json` overrides are worth adding now or later

## References

- [Greptile config docs](https://www.greptile.com/docs/code-review/greptile-config)
- [Greptile config field reference](https://www.greptile.com/docs/code-review/greptile-config-reference)

### CODE REVIEW

**Review by JanuszL — APPROVED** [review]

File: .greptile/config.json [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+    },
+    {
+      "id": "cuda-restrict",
+      "rule": "Non-aliasing input/output kernel pointers must be `__restrict__`-qualified. If a kernel takes both an input and output pointer of the same element type, both should be `__restrict__` unless aliasing is explicitly intended.",
+      "scope": ["**/*.cu", "**/*.cuh"],
+      "severity": "medium"
+    },
+    {
+      "id": "size-t-pixel-index",
+      "rule": "Pixel/byte index math that can exceed 2^31 must use `size_t` (or `int64_t`) to prevent overflow on large images. Pattern: `size_t idx = static_cast<size_t>(y) * W + x`. Flag `int * int` index arithmetic in code that can run on multi-megapixel images.",
```
Comment (mzient) [inline .greptile/config.json]: I'd recommend using signed type unless we really, really need size_t (e.g. for comparison with vector length). Unsigned indices were a mistake of the 80s.

File: .greptile/config.json [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+
+    {
+      "id": "raii-not-manual-cleanup",
+      "rule": "C resources (FILE*, AVFormatContext*, CUDA streams created locally, etc.) must be wrapped in `unique_ptr` with a custom deleter or guarded with `AtScopeExit` — not cleaned up manually across multiple branches.",
+      "scope": ["dali/**/*.cc", "dali/**/*.h", "dali/**/*.cu", "dali/**/*.cuh"],
+      "severity": "high"
+    },
+    {
+      "id": "assert-vs-enforce",
+      "rule": "`assert` is for internal invariants (compiled out in release). User-facing input validation must use `DALI_ENFORCE` (or `DALI_FAIL`) so failures surface in release builds with actionable messages.",
```
Comment (mzient) [inline .greptile/config.json]: We've actually made some effort to phase out DALI_ENFORCE as it uses just one exception type which is Python-hostile. It's better to throw proper C++ exceptions (std::invalid_argument, std::out_of_range) or out errors that build on those (dali::invalid_key)

File: .greptile/config.json [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+      "id": "container-by-const-ref",
+      "rule": "Function parameters of `std::vector`, `std::string`, `std::map`, etc. must be `const T&` (or sink params taken by value with `std::move`). Plain by-value of large containers silently copies.",
+      "scope": ["**/*.cc", "**/*.h", "**/*.cu", "**/*.cuh"],
+      "severity": "medium"
+    },
+    {
+      "id": "no-virtual-without-polymorphism",
+      "rule": "Don't add `virtual` to a method that no derived class overrides — it's dead vtable weight. Don't make instance data members `const` unless immutability is essential; `const` members make the class non-movable.",
+      "scope": ["dali/**/*.cc", "dali/**/*.h", "dali/**/*.cu", "dali/**/*.cuh"],
+      "severity": "low"
```
Comment (mzient) [inline .greptile/config.json:147]: Split this into two rules.

File: .greptile/config.json [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+    },
+    {
+      "id": "test-set-seed",
+      "rule": "Tests using random data must set a deterministic seed (e.g., `np.random.seed(...)` or `pipeline_def(seed=...)`) — otherwise CI flakes hide real bugs.",
+      "scope": ["**/test_*.py", "dali/test/**"],
+      "severity": "medium"
+    },
+    {
+      "id": "test-no-hard-skip",
+      "rule": "Don't unconditionally skip tests in CI when a feature is unavailable — silent skips mean we stop noticing when the feature breaks. Either run the test conditionally with logging, or fail loudly when the prerequisite is missing.",
```
Comment (mzient) [inline .greptile/config.json]: I disagree with this one. It makes running tests in a limited environment unnecessarily hard. This applies especially to multi-gpu tests. We have dedicated jobs for those, but if we're running locally, it doesn't make sense to fail over having just one GPU.

File: .greptile/config.json [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+      "severity": "low"
+    },
+    {
+      "id": "fix-typos",
+      "rule": "Fix typos in identifiers, comments, error messages, and docs. Common ones spotted in DALI history: `sequenece`, `Allready`, `Whtether`, `milliseoncds`, `on-pair`, `Succesful`, `poitner`, `funciton`, `syncrhonized`, `deleteion`, `retruned`, `worskpace`, `explnation`, `dimention`, `indefinately`. Typos in enum values are real bugs (wrong case matched).",
+      "severity": "low"
+    },
+    {
+      "id": "cpp-naming-no-camelcase",
+      "rule": "C++ in DALI uses PascalCase for methods/functions and snake_case for variables — never camelCase. Acronyms in names follow PascalCase too: `StftGpu`, not `STFTGPU`.",
```
Comment (mzient) [inline .greptile/config.json]: Well, _actually_ GPU and CPU are already spelled in all-caps almost everywhere.

File: .greptile/rules.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+- Coverage of `batch_size=1`, empty input, boundary sizes, both CPU and GPU variants.
+- Deterministic seeds (no flaky randomness).
+- `assert_raises(..., glob=...)` with a message pattern, not just exception type.
+- No silent skips when a feature is unavailable in CI.
+
+## Performance review specifics
+
+DALI's hot paths get exercised at thousands of frames per second. Watch for:
+- Unnecessary `cudaStreamSynchronize` calls — every one needs justification.
+- Default-stream launches (`<<<g, b>>>(...)` without an explicit stream) — they create global sync points.
```
Comment (mzient) [inline .greptile/rules.md]: They don't cause synchronization, since we (and almost everyone else) use non-blocking custom streams. They are almost certainly incorrect and _missing_ the necessary synchronization with the actual stream we'd like to use.

File: .greptile/rules.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+- Per-sample work that could be hoisted to constructor or `Setup` (file opens, decoder construction, attribute lookups).
```
Comment (mzient) [inline .greptile/rules.md]: 1. What does per-sample mean in this context?
2. `Setup` has no benefit over `Run`.

File: .greptile/config.json (hunk 1 above)
Comment (jantonguirao) [reply .greptile/config.json]: Updated. Renamed the rule to `int64-pixel-index` and reworded it to prefer signed `int64_t` for index math, only reaching for `size_t` when comparing against `std::vector::size()` and friends.

File: .greptile/config.json (hunk 2 above)
Comment (jantonguirao) [reply .greptile/config.json]: Updated. Renamed to `assert-vs-exception` and reworded to recommend specific C++ exceptions (`std::invalid_argument`, `std::out_of_range`, `std::runtime_error`) or DALI errors built on them (e.g., `dali::invalid_key`). Also added a note that new code should avoid `DALI_ENFORCE`/`DALI_FAIL` since they collapse to a single exception type and are being phased out.

File: .greptile/config.json (hunk 3 above)
Comment (jantonguirao) [reply .greptile/config.json:147]: Done. Split into `no-virtual-without-polymorphism` (drop `virtual`/`override` when nothing overrides) and a new `no-const-data-members` (don't make instance members `const` unless immutability is essential, since it kills value semantics).

File: .greptile/config.json (hunk 4 above)
Comment (jantonguirao) [reply .greptile/config.json]: Fair point — removed the `test-no-hard-skip` rule. Multi-GPU and other hardware-conditional tests have dedicated CI jobs anyway, so failing locally over a missing prerequisite would just be friction.

File: .greptile/config.json (hunk 5 above)
Comment (jantonguirao) [reply .greptile/config.json]: Right, the example was wrong. Reworded to acknowledge that common acronyms (`GPU`, `CPU`, `IO`, `RNG`) stay in all-caps inside identifiers — `StorageCPU`, `InListGPU`, `RunGPU` — and only less-common acronyms (`Stft`, `Jpeg`) follow PascalCase.

File: .greptile/rules.md (hunk 6 above)
Comment (jantonguirao) [reply .greptile/rules.md]: Good catch — reworded. The bullet now reads: a default-stream launch is almost always *missing* the explicit synchronization with the operator's intended stream (since DALI's custom streams are non-blocking), not helpfully syncing with it.

File: .greptile/rules.md (hunk 7 above)
Comment (jantonguirao) [reply .greptile/rules.md]: Reworded both. Dropped the `Setup` reference and the ambiguous "per-sample" framing; the bullet now describes one-time work (file opens, decoder construction, attribute lookups) being repeated every batch and that should be hoisted to the operator constructor.

File: .greptile/config.json [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+      "rule": "Tests must compare against expected values or a reference implementation (NumPy/OpenCV/torchvision), not just check shape/dtype. Ask: \"if the operator returned all zeros, would this test still pass?\" If yes, the assertion is insufficient.",
+      "scope": ["**/test_*.py", "dali/test/**", "dali/**/*_test.cc", "dali/**/*_test.cu"],
+      "severity": "high"
+    },
+    {
+      "id": "test-edge-cases",
+      "rule": "New operator tests must cover: empty input, single sample (`batch_size=1`), boundary values, CPU and GPU variants, and invalid-input rejection (via `assert_raises(..., glob=...)` with the expected message pattern).",
+      "scope": ["**/test_*.py", "dali/test/**", "dali/**/*_test.cc", "dali/**/*_test.cu"],
+      "severity": "medium"
+    },
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply .greptile/config.json:254]: Disagreeing on this one. `dali/python/setup.py.in:62` declares `python_requires='>=3.10, <3.15'` and the only listed classifiers are 3.10–3.14, so DALI's own minimum is 3.10. The `python_min_ver="3.8"/"3.9"` entries in `qa/setup_packages.py` are for the dependency matrix (which third-party wheel versions still ship for older Pythons) — they don't apply to DALI itself. Keeping the rule enabled.

File: .greptile/config.json [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+    {
+      "id": "py-list-of-bool-truthiness",
+      "rule": "`if [False, False, False]:` evaluates to `True` because the list is non-empty. Use `if all(xs):` or `if any(xs):` for list-of-bool checks.",
+      "scope": ["**/*.py"],
+      "severity": "high"
+    },
+    {
+      "id": "py-no-assert-for-user-errors",
+      "rule": "`assert` is disabled with `python -O`. Don't use it for user-facing validation. Raise `ValueError`, `RuntimeError`, or `TypeError` with a message that names the parameter and shows expected vs. actual values.",
+      "scope": ["**/*.py"],
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply .greptile/config.json:159]: Fixed. Reworded to make it clear `inline` is the ODR exemption that lets the same definition appear in multiple TUs so the linker can fold them into one symbol — not a one-definition guarantee.

File: .greptile/config.json [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+    },
+    {
+      "id": "cpp-naming-no-camelcase",
+      "rule": "C++ in DALI uses PascalCase for methods/functions and snake_case for variables — never camelCase. Acronyms in names follow PascalCase too: `StftGpu`, not `STFTGPU`.",
+      "scope": ["dali/**/*.cc", "dali/**/*.h", "dali/**/*.cu", "dali/**/*.cuh"],
+      "severity": "low"
+    },
+    {
+      "id": "pointer-spacing",
+      "rule": "DALI pointer-declaration style: `T *ptr` (space before `*`), not `T* ptr`. Applies to references too: `T &ref`.",
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply .greptile/config.json:333]: Pushing back on this one. `.clang-format` has `DerivePointerAlignment: true` (line 48), which means clang-format derives the style from existing usage in each file; `PointerAlignment: Left` is just the fallback. A quick scan of `dali/` shows ~2400 occurrences of `T *ptr` vs ~340 of `T* ptr`, so right-aligned is overwhelmingly the actual convention. Keeping the rule.

File: .greptile/config.json [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+    },
+    {
+      "id": "doc-no-unverified-perf",
+      "rule": "Don't put unverified performance claims in user-visible docs (\"X is the fastest\", \"always faster than Y\"). If you can't guarantee the claim across configurations, use neutral language.",
+      "scope": ["docs/**/*.rst", "docs/**/*.md", "**/*.py"],
+      "severity": "low"
+    },
+    {
+      "id": "doc-rst-backticks",
+      "rule": "RST formatting: single backticks for cross-reference links (``:ref:`name``` or `` `link <url>`_ ``), double backticks for literal/monospace code (``` ``code`` ```). Mixing them up renders incorrectly.",
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply .greptile/config.json]: Reworded. Dropped the inline backtick examples and described the two roles in prose so the guidance reads cleanly without nested-backtick rendering hazards.

File: .greptile/config.json [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+    },
+
+    {
+      "id": "error-message-with-context",
+      "rule": "Error messages must name the affected parameter/operator, show expected vs. actual values, and end as a complete sentence. Not `\"is 5\"` but `\"got 5 channels, expected 1 or 3.\"`. Use `make_string(...)` for C++ concatenation; f-strings in Python.",
+      "severity": "medium"
+    },
+    {
+      "id": "no-silent-fallback",
+      "rule": "Don't silently degrade or skip when an unsupported feature/config is requested. Either raise/`DALI_FAIL`, or `DALI_WARN` loudly with a reason. Silent fallbacks become invisible regressions.",
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply .greptile/config.json]: Applied. Dropped the `DALI_FAIL` example and pointed at `assert-vs-exception` instead so the two rules don't contradict each other.

File: .greptile/config.json [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+    {
+      "id": "use-dali-utilities",
+      "rule": "Reach for existing DALI utilities before reimplementing: `ConvertSat<uint8_t>(...)` over `lrintf(fminf(fmaxf(...)))`; `clamp<T>(...)` over manual `fminf/fmaxf`; `make_string(...)` over `+`-concatenation; `div_ceil(a,b)` over `(a+b-1)/b`. Search for the helper before writing your own.",
+      "scope": ["dali/**/*.cc", "dali/**/*.h", "dali/**/*.cu", "dali/**/*.cuh"],
+      "severity": "medium"
+    },
+    {
+      "id": "no-side-effects-on-input",
+      "rule": "Operator inputs must not be mutated. Two consumers may read the same input in parallel — mutation creates races. Use `ShareData`/views, or copy if you genuinely need a writable buffer.",
+      "scope": ["dali/operators/**", "dali/kernels/**"],
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply .greptile/config.json]: Applied. Added explicit `{.cc,.h,.cu,.cuh}` extensions to both `dali/operators/**` and `dali/kernels/**` so it lines up with the rest of the rules.

**Review by mzient — APPROVED** [review]
